transformer_engine/pytorch/float8_tensor.py

```diff
@@ -454,61 +456,47 @@ def transpose(
               The first dimension to be transposed
         dim1: int, default = 1
               The second dimension to be transposed
-        update_cache: str or bool, default = "reuse_only"
-                      Memoization behavior. Options are
-                      "reuse_only"/`False` (reuse cached value if
-                      available, otherwise calculate transpose without
-                      caching), "force"/`True` (calculate transpose
-                      and cache), "lazy" (reuse cached value if
-                      available, otherwise calculate transpose and
-                      cache if possible). Caching is only supported
-                      for basic 2D transposes and the cache is reset
-                      after any in-place operations.
-
+        cache: bool, default = `False`
+               If `False`, transpose is calculated and returned.
+               If `True`, the transpose value is cached and can
+               be reused without recomputation by setting the
+               `update_cache` argument to `False`.
+        update_cache: bool, default = `True`
+                      Only used if argument `cache` is `True`, ignored otherwise.
+                      If `True`, the tranpose is recomputed and cached.
+                      If `False`, cached transpose is returned.
         """
 
-        # Check caching mode
-        if not isinstance(update_cache, str):
-            update_cache = "force" if update_cache else "reuse_only"
-        if update_cache not in ("force", "reuse_only", "lazy"):
-            raise ValueError(
-                "Supported values for update_cache are "
-                '"force" (True), "reuse_only" (False), "lazy" '
-                f"(got {update_cache})"
-            )
-
         # Handle non-2D transposes
         if -self.dim() <= dim0 < 0:
             dim0 += self.dim()
         if -self.dim() <= dim1 < 0:
             dim1 += self.dim()
         if self.dim() != 2 or dim0 == dim1:
-            if update_cache == "force":
+            if cache:
                 raise ValueError(
                     "Transpose caching is only supported for basic 2D transposes "
                     f"(ndims={self.dim()}, dim0={dim0}, dim1={dim1})"
                 )
             return super().transpose(dim0, dim1)
 
-        # Clear cache if needed
-        if update_cache == "force":
-            self._transpose = None
-
-        # Compute transpose if needed
-        out = self._transpose
-        if out is None:
-            out = Float8Tensor.make_like(
-                self,
-                data=tex.fp8_transpose(
-                    self._data.contiguous(),
-                    self._fp8_dtype,
-                ),
-            )
+        if not cache:
+            return tex.fp8_transpose(self._data, self._fp8_dtype)
```

Comment:
I can see an argument for moving the transpose caching logic fully within TE, but I think it's a bad idea to return raw `uint8` from this function. `Float8Tensor` keeps track of a lot of things to keep things convenient for the user, and now the user will be forced to manually keep track of `fp8_meta`. Also, this is extremely specific for our GEMM-specific use-case, when `Float8Tensor` should ideally be a generic tensor class that can be upstreamed to PyTorch. `x.transpose(0, 1) + y` will now do something completely weird. If there is some edge-case with how the class currently handles `fp8_meta`, we should address that rather than changing this API.

Reply:
I think I understand the reason for these changes and I think I have a solution that allows us to keep the nice API from `Float8Tensor`. The problem with the existing impl is that resetting the transpose cache will deallocate the transpose data. This messes things up since we want to avoid allocating long-lived buffers within a CUDA graph. What if `Float8Tensor` had `_transpose` and `_data_transpose` attrs? `_data_transpose` is a permanent `uint8` tensor that is never deallocated once created (similar to this PR's impl). `_transpose` is a `Float8Tensor` that is created out of `_data_transpose`, but can be set to `None` after any in-place op in `__torch_dispatch__`.

Reply:
Resolving based on offline discussion.